### PR TITLE
Add cublas offload

### PR DESCRIPTION
**Pull request by yelite**
This PR upstreams the cuBLAS offload from https://github.com/octoml/mlc-llm/pull/8. It adds a flag `--use-cublas` to enable cublas offloading on CUDA target.

cc @masahi

### CODE REVIEW

**Comment by junrushao** [comment]
Just curious - does cuBLAS help with performance, compared with existing solutions?

**Comment by yelite** [comment]
Yes, for q0f16, cuBLAS is faster on GEMM while being slightly slower on GEMV, compared to dlight on A10G. I haven't run extensive benchmark though.

```
Time elapsed: encoding 0.02986598014831543 seconds, decoding 0.02873826026916504 secs
Profiling...
======================= Encoding Profiling =======================
Name                                              Time (ms)   Count   Total time (ms)   Pct (%)   Memory (MB)     Bandwidth (GB/s)  Shape
fused_relax_permute_dims_relax_matmul5_cublas     0.3679      32      11.7732           40.90     172.30          457.3380          (22016, 4096), (1, 6, 4096), (1, 6, 22016)
fused_relax_permute_dims_relax_matmul3_cublas     0.2219      32      7.1016            24.67     96.19           423.2659          (12288, 4096), (1, 6, 4096), (1, 6, 12288)
fused_relax_permute_dims_relax_matmul6_cublas     0.1936      32      6.1965            21.52     86.17           434.5816          (4096, 11008), (1, 6, 11008), (1, 6, 4096)
fused_relax_permute_dims_relax_matmul4_cublas     0.0718      32      2.2974            7.98      32.09           436.5567          (4096, 4096), (1, 6, 4096), (1, 6, 4096)
fused_relax_permute_dims_relax_matmul2_cublas     0.5275      1       0.5275            1.83      250.07          462.9144          (32000, 4096), (1, 1, 4096), (1, 1, 32000)
split1                                            0.0076      32      0.2430            0.84      0.28            36.1646           (1, 6, 12288), (1, 6, 4096), (1, 6, 4096), (1, 6, 4096)
fused_relax_nn_attention1_cutlass1                0.0064      32      0.2052            0.71      0.20            29.7389           (1, 6, 32, 128), (1, 6, 32, 128), (1, 6, 32, 128), (8192,), (1, 6, 32,
 128)
fused_rms_norm_cutlass                            0.0029      65      0.1895            0.66      0.10            34.0250           (1, 6, 4096), (4096,), (1, 6, 4096)
add                                               0.0025      64      0.1621            0.56      0.14            54.2295           (1, 6, 4096), (1, 6, 4096), (1, 6, 4096)
fused_split2_silu1_multiply1                      0.0027      32      0.0850            0.30      0.38            138.9805          (1, 6, 22016), (1, 6, 11008)
cast                                              0.0028      1       0.0028            0.01      0.18            63.3435           (1, 1, 32000), (1, 1, 32000)
take1                                             0.0025      1       0.0025            0.01      250.05          97072.3327        (32000, 4096), (6,), (6, 4096)
slice                                             0.0025      1       0.0025            0.01      0.05            21.6407           (1, 6, 4096), (1, 1, 4096)
Total time: 28.7888 ms

======================= Decoding Profiling =======================
Name                                              Time (ms)   Count   Total time (ms)   Pct (%)   Memory (MB)     Bandwidth (GB/s)  Shape
fused_relax_permute_dims_relax_matmul1_cublas     0.3639      32      11.6441           42.03     172.05          461.7408          (22016, 4096), (1, 1, 4096), (1, 1, 22016)
fused_relax_permute_dims_relax_matmul_cublas      0.2057      32      6.5823            23.76     96.03           455.9147          (12288, 4096), (1, 1, 4096), (1, 1, 12288)
fused_relax_permute_dims_relax_matmul_relax_add1_cublas0.1929      32      6.1718            22.28     86.04           435.6315          (4096, 11008), (1, 1, 11008), (1, 1, 4096), (1, 1, 4096)
fused_relax_permute_dims_relax_matmul_relax_add_cublas0.0715      32      2.2869            8.26      32.02           437.5976          (4096, 4096), (1, 1, 4096), (1, 1, 4096), (1, 1, 4096)
fused_relax_permute_dims_relax_matmul2_cublas     0.5275      1       0.5275            1.90      250.07          462.9298          (32000, 4096), (1, 1, 4096), (1, 1, 32000)
fused_relax_nn_attention_cutlass1                 0.0063      32      0.2020            0.73      0.13            20.5509           (1, 1, 32, 128), (1, 7, 32, 128), (1, 7, 32, 128), (8192,), (1, 1, 32,
 128)
fused_rms_norm1_cutlass                           0.0029      65      0.1908            0.69      0.02            7.7971            (1, 1, 4096), (4096,), (1, 1, 4096)
fused_split_silu_multiply                         0.0029      32      0.0916            0.33      0.06            21.4793           (1, 1, 22016), (1, 1, 11008)
cast                                              0.0028      1       0.0028            0.01      0.18            64.4343           (1, 1, 32000), (1, 1, 32000)
take                                              0.0026      1       0.0026            0.01      250.01          92175.2449        (32000, 4096), (1,), (1, 4096)
Total time: 27.7025 ms
```

```
Time elapsed: encoding 0.06117057800292969 seconds, decoding 0.028194904327392578 secs                                                                                                                    
Profiling...                                                                                                                                                                                              
======================= Encoding Profiling =======================                                                                                                                                        
Name                                              Time (ms)   Count   Total time (ms)   Pct (%)   Memory (MB)     Bandwidth (GB/s)  Shape                                                                 
NT_matmul7                                        0.6386      32      20.4359           33.98     172.30          263.4748          (1, 6, 4096), (22016, 4096), (1, 6, 22016)                            
fused_NT_matmul8_add1                             0.5641      32      18.0511           30.01     86.22           149.2631          (1, 6, 11008), (4096, 11008), (1, 6, 4096), (1, 6, 4096)              
NT_matmul5                                        0.4258      32      13.6248           22.65     96.19           220.6164          (1, 6, 4096), (12288, 4096), (1, 6, 12288)                            
fused_NT_matmul6_add1                             0.2127      32      6.8059            11.32     32.14           147.5767          (1, 6, 4096), (4096, 4096), (1, 6, 4096), (1, 6, 4096)
fused_NT_matmul4_cast                             0.5248      1       0.5248            0.87      250.13          465.4880          (1, 1, 4096), (32000, 4096), (1, 1, 32000)
split1                                            0.0066      32      0.2127            0.35      0.28            41.3290           (1, 6, 12288), (1, 6, 4096), (1, 6, 4096), (1, 6, 4096)
fused_relax_nn_attention1_cutlass1                0.0064      32      0.2041            0.34      0.20            29.8980           (1, 6, 32, 128), (1, 6, 32, 128), (1, 6, 32, 128), (8192,), (1, 6, 32,
 128)
fused_rms_norm_cutlass                            0.0029      65      0.1915            0.32      0.10            33.6701           (1, 6, 4096), (4096,), (1, 6, 4096)
fused_split2_silu1_multiply1                      0.0027      32      0.0868            0.14      0.38            136.0084          (1, 6, 22016), (1, 6, 11008)
take1                                             0.0025      1       0.0025            0.00      250.05          97464.8252        (32000, 4096), (6,), (6, 4096)
slice                                             0.0025      1       0.0025            0.00      0.05            21.5820           (1, 6, 4096), (1, 1, 4096)
Total time: 60.1426 ms

======================= Decoding Profiling =======================
Name                                              Time (ms)   Count   Total time (ms)   Pct (%)   Memory (MB)     Bandwidth (GB/s)  Shape
NT_matmul2                                        0.3574      32      11.4356           41.85     172.05          470.1597          (1, 1, 4096), (22016, 4096), (1, 1, 22016)
NT_matmul                                         0.2005      32      6.4149            23.48     96.03           467.8148          (1, 1, 4096), (12288, 4096), (1, 1, 12288)
fused_NT_matmul3_add                              0.1898      32      6.0734            22.23     86.04           442.6889          (1, 1, 11008), (4096, 11008), (1, 1, 4096), (1, 1, 4096)
fused_NT_matmul1_add                              0.0746      32      2.3877            8.74      32.02           419.1238          (1, 1, 4096), (4096, 4096), (1, 1, 4096), (1, 1, 4096)
fused_NT_matmul4_cast                             0.5246      1       0.5246            1.92      250.13          465.6577          (1, 1, 4096), (32000, 4096), (1, 1, 32000)
fused_relax_nn_attention_cutlass1                 0.0063      32      0.2016            0.74      0.13            20.5854           (1, 1, 32, 128), (1, 7, 32, 128), (1, 7, 32, 128), (8192,), (1, 1, 32,
 128)
fused_rms_norm1_cutlass                           0.0030      65      0.1939            0.71      0.02            7.6722            (1, 1, 4096), (4096,), (1, 1, 4096)
fused_split_silu_multiply                         0.0028      32      0.0889            0.33      0.06            22.1390           (1, 1, 22016), (1, 1, 11008)
take                                              0.0025      1       0.0025            0.01      250.01          97582.6683        (32000, 4096), (1,), (1, 4096)
Total time: 27.3231 ms
```

File: mlc_llm/core.py [hunk 1]
```diff
@@ -146,6 +147,14 @@ class BuildArgs:
             "action": "store_true",
         },
     )
+    use_cublas: bool = field(
+        default=False,
+        metadata={
+            "help": ("Offload matmul to cuBLAS when target is CUDA and TVM has been built"
+                     "with cuBLAS enbaled."),
```
Comment (masahi) [inline mlc_llm/core.py:154]: We should document that cublas offload is meant only for q0f16.

File: mlc_llm/core.py [hunk 2]
```diff
@@ -311,22 +320,27 @@ def mod_transform_before_build(
...
-        if not args.no_cutlass_norm:
+        if has_cutlass and not args.no_cutlass_norm:
             patterns += get_patterns_with_prefix("cutlass.layer_norm")
             patterns += get_patterns_with_prefix("cutlass.rms_norm")
 
-        if use_ft_quant:
+        if has_cutlass and use_ft_quant:
             patterns += get_patterns_with_prefix("cutlass.decode_matmul")
 
+        if has_cublas and args.use_cublas:
```
Comment (masahi) [inline mlc_llm/core.py:341]: Enable only when q0f16?

Comment (sunggg) [reply mlc_llm/core.py:341]: +1

Comment (sunggg) [reply mlc_llm/core.py:341]: Can we enable it by default for q0f16?

Comment (yelite) [reply mlc_llm/core.py]: I updated the code to enable cublas offload in q0f16 and q0f32, and changed the flag to `no_cublas`.

File: mlc_llm/core.py (hunk 1 above)
Comment (yelite) [reply mlc_llm/core.py]: Updated

**Review by junrushao — APPROVED** [review]
Thanks for the numbers! Looks really cool for q0f16

**Review by sunggg — APPROVED** [review]

**Review by junrushao — APPROVED** [review]

**Pull request merged**